File: Player/CustomControls/objc/CustomControls/ControlsViewController.m
//
//  ControlsViewController.m
//  CustomControls
//
//  Copyright © 2019 Brightcove, Inc. All rights reserved.
//

#import "ControlsViewController.h"

#import <MediaPlayer/MediaPlayer.h>

// ** Customize these values **
static NSTimeInterval const kViewControllerControlsVisibleDuration = 5.;
static NSTimeInterval const kViewControllerFadeControlsInAnimationDuration = .1;
static NSTimeInterval const kViewControllerFadeControlsOutAnimationDuration = .2;


@interface ControlsViewController ()

@property (nonatomic, weak) AVPlayer *currentPlayer;

@property (nonatomic, weak) IBOutlet UIView *controlsContainer;
@property (nonatomic, weak) IBOutlet UIButton *playPauseButton;
@property (nonatomic, weak) IBOutlet UILabel *playheadLabel;
@property (nonatomic, weak) IBOutlet UISlider *playheadSlider;
@property (nonatomic, weak) IBOutlet UILabel *durationLabel;
@property (nonatomic, weak) IBOutlet UIView *fullscreenButton;
@property (nonatomic, weak) IBOutlet MPVolumeView *externalScreenButton;

@property (nonatomic, strong) NSTimer *controlTimer;
@property (nonatomic, assign, getter=wasPlayingOnSeek) BOOL playingOnSeek;

@end


@implementation ControlsViewController

#pragma mark BCOVDelegatingSessionConsumerDelegate

-(void)viewDidLoad
{
    [super viewDidLoad];

    // Used for hiding and showing the controls.
    UITapGestureRecognizer *tapRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tapDetected:)];
    tapRecognizer.numberOfTapsRequired = 1;
    tapRecognizer.numberOfTouchesRequired = 1;
    tapRecognizer.delegate = self;
    [self.view addGestureRecognizer:tapRecognizer];

    self.externalScreenButton.showsRouteButton = YES;
    self.externalScreenButton.showsVolumeSlider = NO;
}

- (void)didAdvanceToPlaybackSession:(id<BCOVPlaybackSession>)session
{
    self.currentPlayer = session.player;

    // Reset State
    self.playingOnSeek = NO;
    self.playheadLabel.text = [ControlsViewController formatTime:0];
    self.playheadSlider.value = 0.f;

    [self invalidateTimerAndShowControls];
}

- (void)playbackSession:(id<BCOVPlaybackSession>)session didChangeDuration:(NSTimeInterval)duration
{
    self.durationLabel.text = [ControlsViewController formatTime:duration];
}

- (void)playbackSession:(id<BCOVPlaybackSession>)session didProgressTo:(NSTimeInterval)progress
{
    self.playheadLabel.text = [ControlsViewController formatTime:progress];

    NSTimeInterval duration = CMTimeGetSeconds(session.player.currentItem.duration);
    float percent = progress / duration;
    self.playheadSlider.value = isnan(percent) ? 0.0f : percent;
}

- (void)playbackSession:(id<BCOVPlaybackSession>)session didReceiveLifecycleEvent:(BCOVPlaybackSessionLifecycleEvent *)lifecycleEvent
{
    if ([kBCOVPlaybackSessionLifecycleEventPlay isEqualToString:lifecycleEvent.eventType])
    {
        self.playPauseButton.selected = YES;

        [self reestablishTimer];
    }
    else if([kBCOVPlaybackSessionLifecycleEventPause isEqualToString:lifecycleEvent.eventType])
    {
        self.playPauseButton.selected = NO;

        [self invalidateTimerAndShowControls];
    }
}

#pragma mark IBActions

- (IBAction)handlePlayPauseButtonPressed:(UIButton *)sender
{
    if (sender.selected)
    {
        [self.currentPlayer pause];
    }
    else
    {
        [self.currentPlayer play];
    }
}

- (IBAction)handlePlayheadSliderValueChanged:(UISlider *)sender
{
    NSTimeInterval newCurrentTime = sender.value * CMTimeGetSeconds(self.currentPlayer.currentItem.duration);
    self.playheadLabel.text = [ControlsViewController formatTime:newCurrentTime];
}

- (IBAction)handlePlayheadSliderTouchBegin:(UISlider *)sender
{
    self.playingOnSeek = self.playPauseButton.selected;
    [self.currentPlayer pause];
}

- (IBAction)handlePlayheadSliderTouchEnd:(UISlider *)sender
{
    NSTimeInterval newCurrentTime = sender.value * CMTimeGetSeconds(self.currentPlayer.currentItem.duration);
    CMTime seekToTime = CMTimeMakeWithSeconds(newCurrentTime, 600);

    typeof(self) __weak weakSelf = self;

    [self.currentPlayer seekToTime:seekToTime completionHandler:^(BOOL finished) {

        typeof(self) strongSelf = weakSelf;

        if (finished && strongSelf.wasPlayingOnSeek)
        {
            strongSelf.playingOnSeek = NO;
            [strongSelf.currentPlayer play];
        }

    }];
}

- (IBAction)handleFullScreenButtonPressed:(UIButton *)sender
{
    if (sender.isSelected)
    {
        sender.selected = NO;
        [self.delegate handleExitFullScreenButtonPressed];
    }
    else
    {
        sender.selected = YES;
        [self.delegate handleEnterFullScreenButtonPressed];
    }
}

#pragma mark Hide/Show Controls

- (void)fadeControlsIn
{
    [UIView animateWithDuration:kViewControllerFadeControlsInAnimationDuration animations:^{

        [self showControls];

    } completion:^(BOOL finished) {

        if(finished)
        {
            [self reestablishTimer];
        }

    }];
}

- (void)fadeControlsOut
{
    [UIView animateWithDuration:kViewControllerFadeControlsOutAnimationDuration animations:^{

        [self hideControls];

    }];
}

- (void)hideControls
{
    self.controlsContainer.alpha = 0.f;
}

- (void)showControls
{
    self.controlsContainer.alpha = 1.f;
}

- (void)reestablishTimer
{
    [self.controlTimer invalidate];
    self.controlTimer = [NSTimer scheduledTimerWithTimeInterval:kViewControllerControlsVisibleDuration target:self selector:@selector(fadeControlsOut) userInfo:nil repeats:NO];
}

-(BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldReceiveTouch:(UITouch *)touch
{
    // This makes sure that we don't try and hide the controls if someone is pressing any of the buttons
    // or slider.
    if([touch.view isKindOfClass:[UIButton class]] || [touch.view isKindOfClass:[UISlider class]])
    {
        return NO;
    }

    return YES;
}

- (void)tapDetected:(UIGestureRecognizer *)gestureRecognizer
{
    if(self.playPauseButton.isSelected)
    {
        if(self.controlsContainer.alpha == 0.f)
        {
            [self fadeControlsIn];
        }
        else if (self.controlsContainer.alpha == 1.f)
        {
            [self fadeControlsOut];
        }
    }
}

- (void)invalidateTimerAndShowControls
{
    [self.controlTimer invalidate];
    [self showControls];
}

#pragma mark Class Methods

+ (NSString *)formatTime:(NSTimeInterval)timeInterval
{
    static NSNumberFormatter *numberFormatter;
    static dispatch_once_t once;

    dispatch_once(&once, ^{

        numberFormatter = [[NSNumberFormatter alloc] init];
        numberFormatter.paddingCharacter = @"0";
        numberFormatter.minimumIntegerDigits = 2;

    });

    if (isnan(timeInterval) || !isfinite(timeInterval) || timeInterval == 0)
    {
        return @"00:00";
    }

    NSUInteger hours = floor(timeInterval / 60.0f / 60.0f);
    NSUInteger minutes = (NSUInteger)(timeInterval / 60.0f) % 60;
    NSUInteger seconds = (NSUInteger)timeInterval % 60;

    NSString *formattedMinutes = [numberFormatter stringFromNumber:@(minutes)];
    NSString *formattedSeconds = [numberFormatter stringFromNumber:@(seconds)];

    NSString *ret = nil;
    if (hours > 0)
    {
        ret = [NSString stringWithFormat:@"%@:%@:%@", @(hours), formattedMinutes, formattedSeconds];
    }
    else
    {
        ret = [NSString stringWithFormat:@"%@:%@", formattedMinutes, formattedSeconds];
    }
    
    return ret;
}

@end

File: Player/CustomControls/swift/CustomControls/ControlsViewController.swift
//
//  ControlsViewController.swift
//  CustomControls
//
//  Copyright © 2024 Brightcove, Inc. All rights reserved.
//

import UIKit
import BrightcovePlayerSDK


fileprivate struct ControlConstants {
    static let VisibleDuration: TimeInterval = 5.0
    static let AnimateInDuration: TimeInterval = 0.1
    static let AnimateOutDuraton: TimeInterval = 0.2
}


final class ControlsViewController: UIViewController {

    @IBOutlet fileprivate weak var controlsContainer: UIView!
    @IBOutlet fileprivate weak var playPauseButton: UIButton!
    @IBOutlet fileprivate weak var playheadLabel: UILabel!
    @IBOutlet fileprivate weak var playheadSlider: UISlider!
    @IBOutlet fileprivate weak var durationLabel: UILabel!
    @IBOutlet fileprivate weak var fullscreenButton: UIButton!
    @IBOutlet fileprivate weak var externalScreenButton: MPVolumeView!
    @IBOutlet fileprivate weak var closedCaptionButton: UIButton!

    weak var delegate: ControlsViewControllerFullScreenDelegate?
    weak var currentPlayer: AVPlayer?
    weak var playbackController: BCOVPlaybackController?

    var closedCaptionEnabled: Bool = false {
        didSet {
            closedCaptionButton.isEnabled = closedCaptionEnabled
        }
    }

    fileprivate var controlTimer: Timer?
    fileprivate var playingOnSeek: Bool = false

    fileprivate lazy var ccMenuController: ClosedCaptionMenuController = {
        let ccMenuController = ClosedCaptionMenuController(style: .grouped)
        ccMenuController.controlsView = self
        return ccMenuController
    }()

    fileprivate lazy var numberFormatter: NumberFormatter = {
        let formatter = NumberFormatter()
        formatter.paddingCharacter = "0"
        formatter.minimumIntegerDigits = 2
        return formatter
    }()

    override func viewDidLoad() {
        super.viewDidLoad()

        // Used for hiding and showing the controls.
        let tapRecognizer = UITapGestureRecognizer(target: self,
                                                   action: #selector(tapDetected))
        tapRecognizer.numberOfTapsRequired = 1
        tapRecognizer.numberOfTouchesRequired = 1
        tapRecognizer.delegate = self
        view.addGestureRecognizer(tapRecognizer)

        externalScreenButton.showsRouteButton = true
        externalScreenButton.showsVolumeSlider = false

        closedCaptionButton.isEnabled = false
    }

    @objc
    fileprivate func tapDetected() {
        if playPauseButton.isSelected {
            if controlsContainer.alpha == 0.0 {
                fadeControlsIn()
            } else if (controlsContainer.alpha == 1.0) {
                fadeControlsOut()
            }
        }
    }

    fileprivate func fadeControlsIn() {
        UIView.animate(withDuration: ControlConstants.AnimateInDuration) { [self] in
            showControls()
        } completion:{ [self] (finished: Bool) in
            if finished {
                reestablishTimer()
            }
        }
    }

    @objc
    fileprivate func fadeControlsOut() {
        UIView.animate(withDuration: ControlConstants.AnimateOutDuraton) { [self] in
            hideControls()
        }
    }

    fileprivate func reestablishTimer() {
        controlTimer?.invalidate()
        controlTimer = Timer.scheduledTimer(timeInterval: ControlConstants.VisibleDuration,
                                            target: self,
                                            selector: #selector(fadeControlsOut),
                                            userInfo: nil,
                                            repeats: false)
    }

    fileprivate func hideControls() {
        controlsContainer.alpha = 0.0
    }

    fileprivate func showControls() {
        controlsContainer.alpha = 1.0
    }

    fileprivate func invalidateTimerAndShowControls() {
        controlTimer?.invalidate()
        showControls()
    }

    fileprivate func formatTime(timeInterval: TimeInterval) -> String? {
        if timeInterval.isNaN ||
            !timeInterval.isFinite ||
            timeInterval == 0 {
            return "00:00"
        }

        let hours  = floor(timeInterval / 60.0 / 60.0)
        let minutes = (timeInterval / 60).truncatingRemainder(dividingBy: 60)
        let seconds = timeInterval.truncatingRemainder(dividingBy: 60)

        guard let formattedMinutes = numberFormatter.string(from: NSNumber(value: minutes)),
              let formattedSeconds = numberFormatter.string(from: NSNumber(value: seconds)) else {
            return nil
        }

        return (hours > 0 ?
                "\(hours):\(formattedMinutes):\(formattedSeconds)" :
                    "\(formattedMinutes):\(formattedSeconds)")
    }

    @IBAction
    fileprivate func handleFullScreenButtonPressed(_ button: UIButton) {
        if button.isSelected {
            button.isSelected = false
            delegate?.handleExitFullScreenButtonPressed()
        } else {
            button.isSelected = true
            delegate?.handleEnterFullScreenButtonPressed()
        }
    }

    @IBAction
    fileprivate func handlePlayheadSliderTouchEnd(_ slider: UISlider) {
        if let currentTime = currentPlayer?.currentItem {
            let newCurrentTime = Float64(slider.value) * CMTimeGetSeconds(currentTime.duration)
            let seekToTime = CMTimeMakeWithSeconds(newCurrentTime, preferredTimescale: 600)

            playbackController?.seek(to: seekToTime) { [self] (finished: Bool) in
                playingOnSeek = false
                playbackController?.play()
            }
        }
    }

    @IBAction
    fileprivate func handlePlayheadSliderTouchBegin(_ slider: UISlider) {
        playingOnSeek = playPauseButton.isSelected
        playbackController?.pause()
    }

    @IBAction
    fileprivate func handlePlayheadSliderValueChanged(_ slider: UISlider) {
        if let currentTime = currentPlayer?.currentItem {
            let currentTime = Float64(slider.value) * CMTimeGetSeconds(currentTime.duration)
            playheadLabel.text = formatTime(timeInterval: currentTime)
        }
    }

    @IBAction
    fileprivate func handlePlayPauseButtonPressed(_ button: UIButton) {
        if button.isSelected {
            playbackController?.pause()
        } else {
            playbackController?.play()
        }
    }

    @IBAction 
    fileprivate func handleClosedCaptionButtonPressed(_ button: UIButton) {
        let navController = UINavigationController(rootViewController: ccMenuController)
        present(navController, animated: true, completion: nil)
    }
}


// MARK: - BCOVPlaybackSessionConsumer

extension ControlsViewController: BCOVPlaybackSessionConsumer {

    func didAdvance(to session: BCOVPlaybackSession!) {
        currentPlayer = session.player

        // Reset State
        playingOnSeek = false
        playheadLabel.text = formatTime(timeInterval: 0)
        playheadSlider.value = 0.0

        invalidateTimerAndShowControls()
    }

    func playbackSession(_ session: BCOVPlaybackSession!,
                         didChangeDuration duration: TimeInterval) {
        durationLabel.text = formatTime(timeInterval: duration)
    }

    func playbackSession(_ session: BCOVPlaybackSession!,
                         didProgressTo progress: TimeInterval) {
        playheadLabel.text = formatTime(timeInterval: progress)

        guard let currentItem = session.player.currentItem else {
            return
        }

        let duration = CMTimeGetSeconds(currentItem.duration)
        let percent = Float(progress / duration)
        playheadSlider.value = percent.isNaN ? 0.0 : percent
    }

    func playbackSession(_ session: BCOVPlaybackSession!,
                         didReceive lifecycleEvent: BCOVPlaybackSessionLifecycleEvent!) {

        switch lifecycleEvent.eventType {
            case kBCOVPlaybackSessionLifecycleEventPlay:
                playPauseButton?.isSelected = true
                reestablishTimer()
            case kBCOVPlaybackSessionLifecycleEventPause:
                playPauseButton.isSelected = false
                invalidateTimerAndShowControls()
            case kBCOVPlaybackSessionLifecycleEventReady:
                ccMenuController.currentSession = session
            default:
                break
        }
    }
}


// MARK: - UIGestureRecognizerDelegate

extension ControlsViewController: UIGestureRecognizerDelegate {

    func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, 
                           shouldReceive touch: UITouch) -> Bool {
        // This makes sure that we don't try and hide the controls 
        // if someone is pressing any of the buttons or slider.

        guard let view = touch.view else { return true }

        if view.isKind(of: UIButton.classForCoder()) ||
            view.isKind(of: UISlider.classForCoder()) {
            return false
        }

        return true
    }
}


// MARK: - ControlsViewControllerFullScreenDelegate

protocol ControlsViewControllerFullScreenDelegate: AnyObject {
    func handleEnterFullScreenButtonPressed()
    func handleExitFullScreenButtonPressed()
}
